Improve Binder value change handling with bean validation

tsuoanttila opened the pull request:
Changed values are now only applied to the bean when they pass both field and bean validation. Any change that breaks validation will be pending and is attempted to apply when another change event comes.

This patch also makes the order of status change event handler and listener call order.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/9988)
<!-- Reviewable:end -->

Commit:
Fix bean validation not executed with setBean and ValueChange

Commit:
Fix event firing on value change, fix tests

Commit:
Add JavaDoc, fix partial validation for doWriteIfValid

Commit:
Revert partial field validation change, add TODO comment

hesara commented:
Partial review with some comments and questions. Note also that there are conflicts to resolve.

---

Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1040 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuo3b-RVlGAqC3YIy:-KuTuo3b-RVlGAqC3YIz:b-frlmtn) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1040)):*
> ```Java
>     private BEAN bean;
> 
>     private final Set<Binding<BEAN, ?>> bindings = new LinkedHashSet<>();
> ```

Would it make sense to change this to a list?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1054 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuFBxY9r45OVjqKtW:-KuTuFBxY9r45OVjqKtX:b-25i49w) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1054)):*
> ```Java
>     private BinderValidationStatusHandler<BEAN> statusHandler;
> 
>     private Set<Binding<BEAN, ?>> hasChanges = new LinkedHashSet<>();
> ```

Should this be renamed?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1340 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuYvVaTCQLBn1FYZn:-KuTuYvVaTCQLBn1FYZo:b-7pcifj) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1340)):*
> ```Java
>             this.bean = bean;
>             getBindings().forEach(
>                     b -> ((BindingImpl<BEAN, ?, ?>) b).initFieldValue(bean));
> ```

unnecessary cast?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1381 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTub9Gh3GkOgLGkcTo:-KuTub9Gh3GkOgLGkcTp:b-7pcifj) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1381)):*
> ```Java
>         } else {
>             hasChanges.clear();
>             getBindings().forEach(binding -> ((BindingImpl<BEAN, ?, ?>) binding)
> ```

unnecessary cast?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1457 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTvRfpfpJ_vOqajC7A:-KuTvRfpfpJ_vOqajC7B:b-42v9zm) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1457)):*
> ```Java
>     @SuppressWarnings({ "unchecked" })
>     private BinderValidationStatus<BEAN> doWriteIfValid(BEAN bean,
>             Set<Binding<BEAN, ?>> bindings) {
> ```

Could this parameter be a List if also changing the field type?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1958 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTtqDovAktExc9K6fL:-KuTtqDovAktExc9K6fM:b-ut67it) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1958)):*
> ```Java
>      * @return a list of the bindings
>      */
>     protected List<BindingImpl<BEAN, ?, ?>> getBindings() {
> ```

(this API restricts to only use BindingImpl (though also the previous version had the limitation; potentially breaking change that should probably be mentioned in release notes along with the other things that change)

---

*[server/src/test/java/com/vaadin/data/BinderTest.java, line 773 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTsNQ-2lioo9vfayPv:-KuTsNQ-2lioo9vfayPw:b-7805po) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/test/java/com/vaadin/data/BinderTest.java#L773)):*
> ```Java
>         nameField.setValue("Invalid");
> 
>         Assert.assertTrue("First name change not handled", binder.isValid());
> ```

should this be binder.hasChanges()?

---

*[server/src/test/java/com/vaadin/data/BinderTest.java, line 791 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTsZpKdxhyTB3wDNxo:-KuTsZpLn29CU2_om8c-:b-vm3nym) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/test/java/com/vaadin/data/BinderTest.java#L791)):*
> ```Java
> 
>         Assert.assertTrue("With new first name both changes should be saved",
>                 binder.isValid());
> ```

should we check for hasChanges() here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988#-:-KuTvf-gKYrcsWefn2jx:b54ir02)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1072 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU1FCta5A4RylwblIU:-KuU1FCta5A4RylwblIV:b-d7gc6f) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1072)):*
> ```Java
>     /**
>      * Informs the Binder that a value in Binding was changed. This will trigger
>      * writing the bean if using {@link #setBean(Object)}.
> ```

could mention what happens if not using setBean(Object)

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1465 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU0RzElSDCYqCZp7b7:-KuU0RzElSDCYqCZp7b8:bb7462z) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1465)):*
> ```Java
> 
>         // If no validation errors then update bean
>         // TODO: Enable partial write when some fields don't pass
> ```

TODO

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1479 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU2PArr6DFG5xCkoqJ:-KuU2PArr6DFG5xCkoqK:b-hrn72c) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1479)):*
> ```Java
>                 // Bean validator failed, revert values
>                 restoreBeanState(bean, oldValues);
>             } else if (getBean() == null || bean.equals(getBean())) {
> ```

this branch and when it is used and why is not completely clear to me on the first read - I see it somehow corresponds to the old else branch

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1488 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU2k-vGP8WMtT0ostK:-KuU2k-vGP8WMtT0ostL:baabhp3) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1488)):*
> ```Java
>         BinderValidationStatus<BEAN> status = new BinderValidationStatus<>(this,
>                 bindingStatuses, binderResults);
>         getValidationStatusHandler().statusChange(status);
> ```

What if nothing changed in the status? Did we get an event before? Should we get one? How does this change the behavior for different callers?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1500 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU03_RPGR-ohPmFlSs:-KuU03_RPGR-ohPmFlSt:btrsgnj) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1500)):*
> ```Java
>      * @param oldValues
>      *            the old values
>      */
> ```

since tag

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1522 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU05AYv95ON0gAYJLJ:-KuU05AYv95ON0gAYJLK:btrsgnj) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1522)):*
> ```Java
>      * 
>      * @return map from binding to value
>      */
> ```

since tag

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1527 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU0K0ur7W9Qk4Keu0b:-KuU0K0ur7W9Qk4Keu0c:bw984zc) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1527)):*
> ```Java
>             Set<Binding<BEAN, ?>> bindings) {
>         Map<Binding<BEAN, ?>, Object> oldValues = new HashMap<>();
>         bindings.forEach(binding -> oldValues.put(binding,
> ```

Should this be limited to properties with a setter?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1641 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTzze1peQRBjI2zVjF:-KuTzze1peQRBjI2zVjG:btrsgnj) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1641)):*
> ```Java
>      *            {@code true} to fire validation status events; {@code false}
>      *            to not
>      * @return validation status for the binder
> ```

since tag

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1652 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU-EmQDYVS-7cvP1zo:-KuU-EmQDYVS-7cvP1zp:b-uci9sg) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1652)):*
> ```Java
> 
>         BinderValidationStatus<BEAN> validationStatus;
>         if (bindingStatuses.stream().anyMatch(BindingValidationStatus::isError)
> ```

I'd put the faster condition to check first

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1657 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU-cZOMDGp24QqhIXV:-KuU-cZOMDGp24QqhIXW:b-ung42y) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1657)):*
> ```Java
>                     bindingStatuses, Collections.emptyList());
>         } else {
>             Map<Binding<BEAN, ?>, Object> beanState = getBeanState(getBean(),
> ```

These temporary changes need clear warnings both in javadoc and in release notes as now also done with the other approach to validation.
Alternatively, we could use some kind of a generated bytecode proxy/interceptor or such, but this is mostly an internal implementation detail.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

Commit:
JavaDoc updates, code clean up, test fixes

tsuoanttila commented:
Review status: 1 of 3 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1040 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuo3b-RVlGAqC3YIy:-KuU8P-zMvSa3pO0nhrF:bgh1u3n) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1040)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Would it make sense to change this to a list?
</blockquote></details>

Could be a list, does not really matter in this case.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1054 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuFBxY9r45OVjqKtW:-KuU8M9Z80qeTJ8dwYgn:b44f7ki) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1054)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Should this be renamed?
</blockquote></details>

Yes. Name suggestions 

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1072 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU1FCta5A4RylwblIU:-KuUCmvApkSKWD1CmXHN:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1072)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

could mention what happens if not using setBean(Object)
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1340 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuYvVaTCQLBn1FYZn:-KuUCnO4EFvmkORgB9kK:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1340)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

unnecessary cast?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1381 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTub9Gh3GkOgLGkcTo:-KuUCnvfU_p6JjMAw544:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1381)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

unnecessary cast?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1465 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU0RzElSDCYqCZp7b7:-KuU9H0USWX4BqKBz4Bc:blqy96y) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1465)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

TODO
</blockquote></details>

Should be addressed in a follow-up patch/pr

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1479 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU2PArr6DFG5xCkoqJ:-KuU9WKX6gbn__ldRt05:b-xlhhgc) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1479)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

this branch and when it is used and why is not completely clear to me on the first read - I see it somehow corresponds to the old else branch
</blockquote></details>

The latter `bean.equals(getBean())` is to check for a case where there are unsaved changes in the binder with the currently bound bean but the changes are successfully written to another bean with `writeIfValid(anotherBean)`. In this case the original bean still has unsaved changes, even though they were written into some other bean. (This is something I'm not sure if we should even support)

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1488 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU2k-vGP8WMtT0ostK:-KuU9y28A4zW42RpVC87:b-xvfqiy) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1488)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

What if nothing changed in the status? Did we get an event before? Should we get one? How does this change the behavior for different callers?
</blockquote></details>

writeIfValid never sent any event in any case. This leads to problems with validation status not updating when saving (See for example #9953) 

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1500 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU03_RPGR-ohPmFlSs:-KuUF_EJnL9lw3xpLsfQ:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1500)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

since tag
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1522 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU05AYv95ON0gAYJLJ:-KuUD3Q-In1e3HrS5awG:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1522)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

since tag
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1527 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU0K0ur7W9Qk4Keu0b:-KuUD3gJTyiXn54gC7t5:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1527)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Should this be limited to properties with a setter?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1641 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTzze1peQRBjI2zVjF:-KuUD2yeaEqSdq1UtlrL:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1641)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

since tag
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1652 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU-EmQDYVS-7cvP1zo:-KuUFakoiKcDOEZje03q:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1652)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

I'd put the faster condition to check first
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1657 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU-cZOMDGp24QqhIXV:-KuUASg1ApTBtv-mBDvU:b-15k9xk) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1657)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

These temporary changes need clear warnings both in javadoc and in release notes as now also done with the other approach to validation.
Alternatively, we could use some kind of a generated bytecode proxy/interceptor or such, but this is mostly an internal implementation detail.
</blockquote></details>

We had a long discussion back then about using proxy/clone/whatever approach to test temporary changes, but it was dropped as we can run into multiple issues with it. If we had that possibility the whole implementation of the Binder would be rather different.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1958 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTtqDovAktExc9K6fL:-KuUAq_eqhN7IYuD5OpW:berxvj5) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1958)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(this API restricts to only use BindingImpl (though also the previous version had the limitation; potentially breaking change that should probably be mentioned in release notes along with the other things that change)
</blockquote></details>

In follow-up patches/PRs this method should just return the binding set/list with the Binding interface and the Binder should not care what is the actual implementation. End result of all upcoming changes to Binder should be to never access a Binding as BindingImpl.

---

*[server/src/test/java/com/vaadin/data/BinderTest.java, line 773 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTsNQ-2lioo9vfayPv:-KuUEdhXuEatwSabOZPM:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/test/java/com/vaadin/data/BinderTest.java#L773)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

should this be binder.hasChanges()?
</blockquote></details>

Done.

---

*[server/src/test/java/com/vaadin/data/BinderTest.java, line 791 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTsZpKdxhyTB3wDNxo:-KuUEdwRK40cQpBxMa4a:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/test/java/com/vaadin/data/BinderTest.java#L791)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

should we check for hasChanges() here?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 1 of 2 files at r3.
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

hesara commented:
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1054 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuFBxY9r45OVjqKtW:-Ku_8bkTGg4UoqNmC1eq:b4ty9zj) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1054)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

Yes. Name suggestions 
</blockquote></details>

pendingChanges? changedBindings?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1958 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTtqDovAktExc9K6fL:-Ku_8wPTDtjIfdf4hJdL:bk5sdm5) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1958)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

In follow-up patches/PRs this method should just return the binding set/list with the Binding interface and the Binder should not care what is the actual implementation. End result of all upcoming changes to Binder should be to never access a Binding as BindingImpl.
</blockquote></details>

So effectively this should return a Collection<Binding<BEAN, ?, ?>>? Can it already do so? When was getBindings() introduced and would it break things?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

Commit:
Rename variables, change types

tsuoanttila commented:
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1040 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuo3b-RVlGAqC3YIy:-KudNjT632Nv0LCKW1iD:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1040)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

Could be a list, does not really matter in this case.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1054 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTuFBxY9r45OVjqKtW:-KudNiz4cdcH4z-Sfpt0:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1054)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

pendingChanges? changedBindings?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1457 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTvRfpfpJ_vOqajC7A:-KudNkkK6VVTlgHPEEgS:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1457)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Could this parameter be a List if also changing the field type?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1958 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuTtqDovAktExc9K6fL:-KudOidsZnjOy0h254XR:buovwsw) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1958)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

So effectively this should return a Collection<Binding<BEAN, ?, ?>>? Can it already do so? When was getBindings() introduced and would it break things?
</blockquote></details>

Introduced a year ago as protected Set<...>, changed to protected Collection<...>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1479 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU2PArr6DFG5xCkoqJ:-KudRRiC6QKefYVXsZNZ:b-hkyldx) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1479)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

The latter `bean.equals(getBean())` is to check for a case where there are unsaved changes in the binder with the currently bound bean but the changes are successfully written to another bean with `writeIfValid(anotherBean)`. In this case the original bean still has unsaved changes, even though they were written into some other bean. (This is something I'm not sure if we should even support)
</blockquote></details>

could there be comments explaining how it works, and why doing various things

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1073 at r4](https://reviewable.io:443/reviews/vaadin/framework/9988#-KudQlgXsjPMe-nd-lZH:-KudQlgXsjPMe-nd-lZI:bt8h0yl) ([raw file](https://github.com/vaadin/framework/blob/81117594a501fcb9ab5dd0f41b0b38bcfe2305eb/server/src/main/java/com/vaadin/data/Binder.java#L1073)):*
> ```Java
>     /**
>      * Informs the Binder that a value in Binding was changed. This will trigger
>      * writing the bean if using {@link #setBean(Object)}. If using
> ```

not really clear from this that the bean is validated before being written in the first case

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1531 at r4](https://reviewable.io:443/reviews/vaadin/framework/9988#-KudR6kJj-6HNjeoacJ7:-KudR6kJj-6HNjeoacJ8:by0me6d) ([raw file](https://github.com/vaadin/framework/blob/81117594a501fcb9ab5dd0f41b0b38bcfe2305eb/server/src/main/java/com/vaadin/data/Binder.java#L1531)):*
> ```Java
>             Collection<Binding<BEAN, ?>> bindings) {
>         Map<Binding<BEAN, ?>, Object> oldValues = new HashMap<>();
>         bindings.stream().map(binding -> (BindingImpl) binding)
> ```

why not use getBindings() everywhere so that overriding it actually lets you change the bindings - that would also simplify the code here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

Commit:
Improve JavaDoc, add comment

tsuoanttila commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1479 at r2](https://reviewable.io:443/reviews/vaadin/framework/9988#-KuU2PArr6DFG5xCkoqJ:-KudWQi0E2MIw4G-_Nfa:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/e268af9e433665e32d435aeb7fdfed6cf2440293/server/src/main/java/com/vaadin/data/Binder.java#L1479)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

could there be comments explaining how it works, and why doing various things
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1073 at r4](https://reviewable.io:443/reviews/vaadin/framework/9988#-KudQlgXsjPMe-nd-lZH:-KudT2hnroPFbA5vJuIG:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/81117594a501fcb9ab5dd0f41b0b38bcfe2305eb/server/src/main/java/com/vaadin/data/Binder.java#L1073)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

not really clear from this that the bean is validated before being written in the first case
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1531 at r4](https://reviewable.io:443/reviews/vaadin/framework/9988#-KudR6kJj-6HNjeoacJ7:-KudSELWnLwn151EN_Go:bofm9r6) ([raw file](https://github.com/vaadin/framework/blob/81117594a501fcb9ab5dd0f41b0b38bcfe2305eb/server/src/main/java/com/vaadin/data/Binder.java#L1531)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

why not use getBindings() everywhere so that overriding it actually lets you change the bindings - that would also simplify the code here
</blockquote></details>

The passed collection of bindings either contains everything or only the changed ones. In the case of trying to write 2 changed bindings out of total 200 bindings, there is a lot less method calls. Can use getBindings as well, this is just a minor optimisation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

tsuoanttila committed:
Merge branch 'master' into beanvalidation_simplified

hesara commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 2015 at r6](https://reviewable.io:443/reviews/vaadin/framework/9988#-KudZM9X4yZegJBKU29X:-KudZM9X4yZegJBKU29Y:bf60jkw) ([raw file](https://github.com/vaadin/framework/blob/148cfe729e2449736e3d5259b0e350e1e004f417/server/src/main/java/com/vaadin/data/Binder.java#L2015)):*
> ```Java
>      * @return a list of the bindings
>      */
>     protected Collection<BindingImpl<BEAN, ?, ?>> getBindings() {
> ```

mention API change(s) in release notes

---

*[server/src/test/java/com/vaadin/data/BinderTest.java, line 899 at r6](https://reviewable.io:443/reviews/vaadin/framework/9988#-KudZBoOhXLO7LCb1ydf:-KudZBoOhXLO7LCb1ydg:bl95yw1) ([raw file](https://github.com/vaadin/framework/blob/148cfe729e2449736e3d5259b0e350e1e004f417/server/src/test/java/com/vaadin/data/BinderTest.java#L899)):*
> ```Java
>                 bindingHandler.callCount);
>     }
>       
> ```

(extra whitespace here and below)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

Commit:
Fix test to account for events fired by doWriteIfValid

hesara commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->

Commit:
Add mention of changed return type in Binder

hesara commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9988)*
<!-- Sent from Reviewable.io -->
